src/parallel.cpp

```diff
@@ -17,38 +17,49 @@
 
 #include <boost/thread/thread.hpp>
 
-#define MAX_SCRIPT_CHECK_QUEUES 4
-
-static CCheckQueue<CScriptCheck> scriptcheckqueue1(128);
-static CCheckQueue<CScriptCheck> scriptcheckqueue2(128);
-static CCheckQueue<CScriptCheck> scriptcheckqueue3(128);
-static CCheckQueue<CScriptCheck> scriptcheckqueue4(128);
-
 using namespace std;
 
+static const unsigned int nScriptCheckQueues = 4;
+std::unique_ptr<CParallelValidation> PV;
 
-void AddScriptCheckThreads(int i, CCheckQueue<CScriptCheck> *pqueue)
+static void HandleBlockMessageThread(CNode *pfrom, const std::string strCommand, const CBlock block, const CInv inv);
+
+static void AddScriptCheckThreads(int i, CCheckQueue<CScriptCheck> *pqueue)
 {
     ostringstream tName;
     tName << "bitcoin-scriptchk" << i;
     RenameThread(tName.str().c_str());
     pqueue->Thread();
 }
 
-void AddAllScriptCheckQueuesAndThreads(int nScriptCheckThreads, boost::thread_group *threadGroup)
+CParallelValidation::CParallelValidation(int threadCount, boost::thread_group *threadGroup)
+    : semaphore(nScriptCheckQueues)
 {
-    CCheckQueue<CScriptCheck> *pqueue[MAX_SCRIPT_CHECK_QUEUES] = {
-        &scriptcheckqueue1, &scriptcheckqueue2, &scriptcheckqueue3, &scriptcheckqueue4};
+    if (threadCount <= 1)
+        threadCount = 0;
+    else if (threadCount > MAX_SCRIPTCHECK_THREADS)
+        threadCount = MAX_SCRIPTCHECK_THREADS;
+    nThreads = threadCount;
+
+    LogPrintf("Using %d threads for script verification\n", threadCount);
 
-    for (int i = 0; i < MAX_SCRIPT_CHECK_QUEUES; i++)
+    while (QueueCount() < nScriptCheckQueues)
     {
-        allScriptCheckQueues.Add(pqueue[i]);
-        for (int j = 0; j < nScriptCheckThreads; j++)
-            threadGroup->create_thread(boost::bind(&AddScriptCheckThreads, i + 1, pqueue[i]));
+        std::unique_ptr<CCheckQueue<CScriptCheck> > queue(new CCheckQueue<CScriptCheck>(128));
+
+        for (unsigned int i = 0; i < nThreads; i++)
+            threadGroup->create_thread(boost::bind(&AddScriptCheckThreads, i + 1, queue.get()));
+
+        vQueues.push_back(std::move(queue));
```

Comment:
@kyuupichan please use the minimum effective std code.  In this simple, constrained function we do not need help tracking a pointer.

There are significant reasons not to use unique_ptr.
1. readability:  To understand this code, you need to understand details of some relatively obscure std classes and functions: unique_ptr, unique_ptr.get, and std::move.  Compare that to understanding a pointer.
2. possibility of incorrect behavior:  when unique_ptr falls out of scope it deletes the object referenced.  But in this case we never want that to happen.

Do you disagree?


Finally, by my read of the spec, this code is incorrect.  When you call queue.get() you expose the pointer of the queue into the created thread.

Yet, if you read the details of std::move():

"If the argument identifies a resource-owning object, these overloads have the option, but aren't required, to move any resources held by the argument. For example, a move constructor of a linked list might copy the pointer to the head of the list and store nullptr in the argument instead of allocating and copying individual nodes."

I read this as std::move may take over the pointer, or might copy it depending on implementation.  If it copied, the thread would be pointing to deleted data once unique_ptr fell out of scope (looks like this std implementation is probably taking the pointer over, or we'd be getting a seg fault, but that does not mean all will).

Even if my reading is incorrect, there's a confusion and complexity there that disappears if you just used a pointer.

If you want PRs merged faster, please don't use the Formlabs SLA 3d printer when you need a checker.  A bottle cap works fine.